ethcore/src/blockchain/blockchain.rs

```diff
@@ -843,6 +832,80 @@ impl BlockChain {
 		}
 	}
 
+	/// Update the best ancient block to the given hash, after checking that
+	/// it's directly linked to the currently known best ancient block
+	pub fn update_best_ancient_block(&self, hash: &H256) {
+		// Get the block view of the next ancient block (it must
+		// be in DB at this point)
+		let block_view = match self.block(hash) {
+			Some(v) => v,
+			None => return,
+		};
+
+		// So that `best_ancient_block` gets unlocked before calling
+		// `set_best_ancient_block`
+		{
+			// Get the target hash ; if there are no ancient block,
+			// it means that the chain is already fully linked
+			let best_ancient_block = self.best_ancient_block.read();
+			let cur_ancient_block = match *best_ancient_block {
+				Some(ref b) => b,
+				None => return,
+			};
+			let target_hash = cur_ancient_block.hash;
+
+			// Ensure that the new best ancient block is after the current one
+			if block_view.number() <= cur_ancient_block.number {
+				return;
+			}
+
+			let mut block_hash = *hash;
+			let mut is_linked = false;
+
+			loop {
+				if block_hash == target_hash {
+					is_linked = true;
+					break;
+				}
+
+				match self.block_details(&block_hash) {
+					Some(block_details) => {
+						block_hash = block_details.parent;
+					},
+					None => break,
+				}
+			}
+
+			if !is_linked {
+				trace!(target: "blockchain", "The given block {:x} is not linked to the known ancient block {:x}", hash, target_hash);
+				return;
+			}
+		}
+
+		let mut batch = self.db.key_value().transaction();
+		self.set_best_ancient_block(block_view.number(), hash, &mut batch);
+		self.db.key_value().write(batch).expect("Low level database error.");
+	}
+
+	/// Set the best ancient block with the given value: private method
+	/// `best_ancient_block` must not be locked, otherwise a DeadLock would occur
+	fn set_best_ancient_block(&self, block_number: BlockNumber, block_hash: &H256, batch: &mut DBTransaction) {
+		let mut best_ancient_block = self.best_ancient_block.write();
+		let ancient_number = best_ancient_block.as_ref().map_or(0, |b| b.number);
+		if self.block_hash(block_number + 1).is_some() {
```

Comment:
I see that this line hasn't been changed, but I believe there is a logic error here. Subsequent calls to `insert_unordered_block` store block hashes only in the pending collection. Therefore it may happen that `best_ancient_block` will be never set to `Some` and `self.block_hash()` will always return `None` leading the database into unknown state (maybe it gets stack exactly because of that?)

Reply:
I think it's fine because the block hash is written to DB before calling `set_best_ancient_block` in `prepare_update`.